devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy

```diff
@@ -90,33 +90,66 @@ def parse(String description) {
             List<Map> descMaps = collectAttributes(descMap)
             def liftmap = descMaps.find { it.attrInt == ATTRIBUTE_POSITION_LIFT }
             if (liftmap) {
-                state.level = Integer.parseInt(liftmap.value, 16)
                 if (liftmap.value == "64") { //open
-                    sendEvent(name: "windowShade", value: "open")
                     sendEvent(name: "level", value: "100")
+                    levelEventHandler(100)
                 } else if (liftmap.value == "00") { //closed
-                    sendEvent(name: "windowShade", value: "closed")
                     sendEvent(name: "level", value: "0")
+                    levelEventHandler(0)
                 } else {
-                    sendEvent(name: "windowShade", value: "partially open")
                     sendEvent(name: "level", value: zigbee.convertHexToInt(liftmap.value))
+                    levelEventHandler(zigbee.convertHexToInt(liftmap.value))
                 }
+                state.level = Integer.parseInt(liftmap.value, 16)
```

Comment:
Wouldn't it make more sense to consolidate all of this?

```
            if (liftmap) {
                def newLevel = zigbee.convertHexToInt(liftmap.value)
                sendEvent(name: "level", value: newLevel)
                levelEventHandler(newLevel)
                state.level = newLevel
            }
```

Or, don't use state.level and instead use `device.currentValue("level")` below where you had state.level, and send the level event in levelEventHandler after you have checked the previous level.

```
            if (liftmap && liftmap.value) {
                levelEventHandler(zigbee.convertHexToInt(liftmap.value))
            }

...

def levelEventHandler(currentLevel) {
    def lastLevel = device.currentValue("level")

    // checks for level logic
    // sendEvent for new windowShade state
    sendEvent(name: "level", value: currentLevel)
}
```

This will reduce the risk that state.level is ever out of sync, or ever gets updated before you want it to.